[LWZ19] iP

LWZ19 opened the pull request:
### **DukePro**

> "Your mind is for having ideas, not holding them." – David Allen ([source](https://dansilvestre.com/productivity-quotes)) 

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ _SUPER FAST_ to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
___
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

LWZ19 committed:
Level 0. Rename, Greet, Exit

LWZ19 committed:
Level 1. Echo

LWZ19 committed:
Level 2. Add, List

LWZ19 committed:
Level 3. Mark as Done

LWZ19 committed:
Level 4. ToDos, Events, Deadlines

LWZ19 committed:
A-TextUiTesting. Automated Text UI Testing

LWZ19 committed:
Level 5. Handle Errors

LWZ19 committed:
Level 6. Delete

LWZ19 committed:
A-Enums. Use Enums

LWZ19 committed:
Made the code more OOP

LWZ19 committed:
Create Storage class to access files

LWZ19 committed:
Level 7. Save

LWZ19 committed:
Merge branch 'branch-Level-7'

LWZ19 committed:
Add minimal goal of accept date and print data.

LWZ19 committed:
Level 8. Dates and Times

LWZ19 committed:
Merge branch 'branch-Level-8'

LWZ19 committed:
Add Ui class

LWZ19 committed:
A-MoreOOP. Use More OOP

LWZ19 committed:
Merge branch 'branch-A-MoreOOP'

LWZ19 committed:
Put all classes in duke package

LWZ19 committed:
A-Packages. Organize into Packages

LWZ19 committed:
Merge branch 'branch-A-Packages'

LWZ19 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

LWZ19 committed:
Set up gradle.

LWZ19 committed:
A-Gradle. Use Gradle

LWZ19 committed:
Merge branch 'branch-A-Gradle'

LWZ19 committed:
A-JUnit. Add JUnit Tests

LWZ19 committed:
Merge branch 'branch-A-JUnit'

LWZ19 committed:
Fix minor bug on mark task

LWZ19 committed:
Add javaDoc to some classes, and edit the output format

LWZ19 committed:
Reorder import statments

LWZ19 committed:
Add command find, to find tasks

LWZ19 committed:
Add JUnit test for find command

LWZ19 committed:
Merge branch 'branch-A-CodingStandard'

LWZ19 committed:
Merge branch 'branch-Level-9'

LWZ19 committed:
Add javaDoc to all classes

LWZ19 committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java

LWZ19 committed:
Improve UI display

LWZ19 committed:
Remove ANSI code

geraldngjx commented on src/main/java/duke/Keyword.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Represents the keywords that the chatbot can understand.
+ */
+public enum Keyword {
```
I like how you have abstracted every command into their individual keyword category using enums!

geraldngjx commented on src/main/java/duke/Time.java:
```diff
@@ -0,0 +1,62 @@
+ * Time class is used to parse the date and time from
+ * {@code String} to {@code LocalDateTime} or {@code LocalDate} object.
+ */
+public class Time {
```
I like how you have created a separate class for Time in order to handle formatting issues!

geraldngjx commented on src/test/java/duke/StorageStub.java:
```diff
@@ -0,0 +1,34 @@
+ * A <code>StorageStub</code> object corresponds to a storage
+ * that does not actually write to a file. It is used for testing.
+ */
+public class StorageStub extends Storage {
```
Good effort in creating a StorageStub specially for testing purposes!

geraldngjx reviewed (commented):
Impressive work on maintaining a consistent coding style and using descriptive variable names! It's clear that you've put thought into making the codebase clean and comprehensible. Well-named variables greatly enhance the code's readability, and your attention to detail in this aspect is commendable. Keep up the excellent work!

geraldngjx commented on src/main/java/duke/exception/ManipulateException.java:
```diff
@@ -0,0 +1,21 @@
+ * Represents an exception thrown when the user enters an invalid
+ * manipulate command keyword.
+ */
+public class ManipulateException extends DukeException {
```
Perhaps you can include more explanation on what does "manipulate command keyword" mean?

woojiahao replied:
Seconded!

geraldngjx commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,42 @@
+ * Represents a command to be executed by the chatbot. A <code>Command</code> object
+ * corresponds to an executable command.
+ */
+public abstract class Command {
```
I like how you have created an abstract class Command in order to handle all existing commands to enhance code readability!

geraldngjx reviewed (commented):
Overall, great job! I added a few more comments and you might consider adding a few more lines of explanation for certain terms (e.g. manipulate). Other than that, it is an excellent piece of work!

woojiahao commented on src/main/java/duke/Keyword.java:
```diff
@@ -0,0 +1,27 @@
+    EVENT("event"),
+    DEADLINE("deadline"),
+    PRINT_DATE("print_date"),
+    FIND("find");
```
Amazing use of enums and enum constructor values

woojiahao commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,136 @@
+ * <p><br>Commands are in the form: {@code <keyword> <command body>}</p><br>
+ * <p>Where the command body is optional. The keyword is case-insensitive.
+ * The command body is the rest of the user input after the keyword.
+ * If the command body is present, it must be separated from the keyword by a space.</p>
```
Great use of HTML in Javadocs, perhaps you can consider adding actual line breaks within the Javadocs so that it is readable at a glance too?

woojiahao commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,136 @@
+        case MARK:
+        case UNMARK:
+        case DELETE:
+            throw new ManipulateException(err, key.getKeyword());
```
I believe you should include a `// Fallthrough` comment as per the coding standard: https://se-education.org/guides/conventions/java/intermediate.html, other than that, great way to use fallthrough!

woojiahao commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    /**
+     * Load the task list from the file.
```
Would it be better if the Javadocs start with `-s`? Just a food for thought!

woojiahao commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,192 @@
+     */
+    public void markTask(int index, Keyword key, Ui ui) throws DukeException {
+        switch (key) {
+        case MARK:
+        case UNMARK:
```
Would be great if you could add an explicit fallthrough comment!

woojiahao commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,192 @@
+        if (taskList.isEmpty()) {
+            throw new DukeException("OOPS!!! There is nothing in the list, yet!");
+        }
+        List<String> tasksFound = new ArrayList<>();
+        List<String> taskIndexFound = new ArrayList<>();
```
Maybe you could phrase it as foundTasks instead? Just a thought

woojiahao commented on src/main/java/duke/Time.java:
```diff
@@ -0,0 +1,62 @@
+ */
+public class Time {
+
+    private static final String DATE_FORMAT = "d/M/yyyy";
+    private static final String TIME_FORMAT = "HH:mm";
+    private static final String DATE_DISPLAY_FORMAT = "d MMM yyyy";
+    private static final String TIME_DISPLAY_FORMAT = "h:mma";
+    public static final DateTimeFormatter DATE_FORMATTER =
+            DateTimeFormatter.ofPattern(DATE_FORMAT);
+    public static final DateTimeFormatter DATE_TIME_FORMATTER =
+            DateTimeFormatter.ofPattern(DATE_FORMAT + " " + TIME_FORMAT);
+    public static final DateTimeFormatter DATE_TIME_DISPLAY_FORMATTER =
+            DateTimeFormatter.ofPattern(DATE_DISPLAY_FORMAT + ", " + TIME_DISPLAY_FORMAT);
+    public static final DateTimeFormatter DATE_DISPLAY_FORMATTER =
+            DateTimeFormatter.ofPattern(DATE_DISPLAY_FORMAT);
```
Great that you have a separation of the formatting so that you can easily compose new formats

woojiahao commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,202 @@
+
+    protected static final String INDENT = "     ";
+    protected static final String NAME = "404";
+    private static final String NAME_ART =
+              "               _                _               _                      \n"
+            + "           _  /\\ \\            / /\\          _  /\\ \\               \n"
+            + "          /\\_\\\\ \\ \\          / /  \\        /\\_\\\\ \\ \\        \n"
+            + "         / / / \\ \\ \\        / / /\\ \\      / / / \\ \\ \\          \n"
+            + "        / / /   \\ \\ \\      / / /\\ \\ \\    / / /   \\ \\ \\        \n"
+            + "        \\ \\ \\____\\ \\ \\    /_/ /  \\ \\ \\   \\ \\ \\____\\ \\ \\ \n"
+            + "         \\ \\________\\ \\   \\ \\ \\   \\ \\ \\   \\ \\________\\ \\ \n"
+            + "          \\/________/\\ \\   \\ \\ \\   \\ \\ \\   \\/________/\\ \\  \n"
+            + "                    \\ \\ \\   \\ \\ \\___\\ \\ \\            \\ \\ \\ \n"
+            + "                     \\ \\_\\   \\ \\/____\\ \\ \\            \\ \\_\\ \n"
+            + "                      \\/_/    \\_________\\/             \\/_/";
```
Very creative name art

woojiahao commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,42 @@
+     *
+     * @return Whether the command is an exit command.
+     */
+    public boolean isExit() {
+        return this.isExit;
```
Perhaps it's better to not name the getter method after the variable name?

woojiahao commented on src/main/java/duke/exception/EventException.java:
```diff
@@ -0,0 +1,19 @@
+/**
+ * Represents an exception thrown when the user enters an invalid event command.
+ */
+public class EventException extends DukeException {
```
Great idea to use dedicated errors to print different invalid inputs!

woojiahao commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,68 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDateTime by;
```
Just curious but is there a reason why this is a protected field?

woojiahao reviewed (commented):
Overall great work! I think there are very small coding standard nits that have been pointed out. I would also like to point out that perhaps it is better if you had a separation of concerns with the `TaskList` and the printing of each action. Since you already have a set of `Command` classes, it might be better to explore moving the printing behavior to each `Command` subclass so that the `TaskList` can focus on just manipulating the `List<Task>`! Just a food for thought! Great job otherwise, keep it up!

lsyurea commented on src/main/java/duke/Keyword.java:
```diff
@@ -0,0 +1,27 @@
+    LIST("list"),
+    MARK("mark"),
+    UNMARK("unmark"),
+    DELETE("delete"),
```
Love your use of enums

lsyurea commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,136 @@
+     * @return The corresponding command.
+     * @throws DukeException If the user input is invalid.
+     */
+    public static Command parse(String fullCommand) throws DukeException {
```
Love your use of Command classes in parse

lsyurea commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,42 @@
+ * corresponds to an executable command.
+ */
+public abstract class Command {
+    private boolean isExit;
```
Love that your isExit sounds like a boolean

lsyurea commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,24 @@
+ * Represents a checked exception thrown when the chatbot
+ * encountered an error state.
+ */
+public class DukeException extends Exception {
```
Love that you have Duke exception with multiple classes to handle the different exceptions

lsyurea reviewed (commented):
Overall well Done

LWZ19 committed:
Add GUI

LWZ19 committed:
Improve GUI and add varargs

LWZ19 committed:
Merge branch 'branch-Level-10'

LWZ19 committed:
Add assertions to all classes

There are no checks for assumptions in all the classes.

No checks for assumptions means that the code might fail, if the
programmer accidently check some code or forget about the assumptions.

Let's add checks for important assumptions as assertions in all the
classes.

LWZ19 committed:
Improve code quality

The code have various violations.

Better code quality can have better readability and it can be more
easily understood by programmers and easier to debug.

Let's,
* Abstract out long method and deep nested code
* Avoid complicated expressions

LWZ19 committed:
Add stream to Ui and TaskList class

Using for loops.

Stream can be used to replace for loops for better readability and
maintenance.

Let's change all the for loop in Ui and TaskList to streams.

LWZ19 committed:
Merge pull request #1 from LWZ19/branch-A-Assertions

Add assertions to all classes

LWZ19 committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java

LWZ19 committed:
Merge branch 'branch-A-Streams'

# Conflicts:
#	src/main/java/duke/util/TaskList.java

LWZ19 committed:
Set up CI

LWZ19 committed:
Fix code style issues

LWZ19 committed:
Add C-DetectDuplicates extension

LWZ19 committed:
Merge branch 'branch-C-DetectDuplicates'

LWZ19 committed:
Simplify Storage class

LWZ19 committed:
Add C-FlexibleDataSource

LWZ19 committed:
Merge branch 'branch-C-FlexibleDataSource'

LWZ19 committed:
Fix GUI issues

LWZ19 committed:
Add C-Sort

LWZ19 committed:
Merge branch 'branch-C-Sort'

LWZ19 committed:
Add alias feature

LWZ19 committed:
Add Javadoc to Alias class

LWZ19 committed:
Improve GUI

LWZ19 committed:
Add asymmetric GUI for the bot

LWZ19 committed:
Add C-Help extension, and improve personality and GUI

LWZ19 committed:
Add error handling of start date is later then end date

LWZ19 committed:
Update User Guide

LWZ19 committed:
Fix style

LWZ19 committed:
Minor UG fix

LWZ19 committed:
Add section links

LWZ19 committed:
Fix section link bug

LWZ19 committed:
Fix read file bug

LWZ19 committed:
Add more Junit tests

LWZ19 committed:
Add Ui image

LWZ19 committed:
Fix folder not found error

LWZ19 committed:
Consistent format

LWZ19 committed:
Fix style in MainWindow.fxml
